cmd/mt-update-ttl/main.go

```diff
@@ -129,38 +143,77 @@ func getTTL(now, ts, ttl int) int {
-		} else {
-			query = fmt.Sprintf("INSERT INTO %s (data, key, ts) values(?,?,?) USING TTL %d", tableOut, newTTL)
-		}
-		if *verbose {
-			fmt.Printf("processing rownum=%d table=%q key=%q ts=%d data='%x'\n", rownum, tableIn, key, ts, data)
-			fmt.Println("Query:", query)
-
+	queryTpl := fmt.Sprintf("SELECT token(key), ts, data FROM %s where key=? AND ts>=? AND ts<?", tableIn)
+
+	for key := range jobs {
+		iter := session.Query(queryTpl, key, startTime, endTime).Iter()
+		for iter.Scan(&token, &ts, &data) {
+			newTTL := getTTL(int(time.Now().Unix()), ts, ttl)
+			if tableIn == tableOut {
+				query = fmt.Sprintf("UPDATE %s USING TTL %d SET data = ? WHERE key = ? AND ts = ?", tableIn, newTTL)
+			} else {
+				query = fmt.Sprintf("INSERT INTO %s (data, key, ts) values(?,?,?) USING TTL %d", tableOut, newTTL)
+			}
+			if *verbose {
+				log.Printf("processing rownum=%d table=%q key=%q ts=%d data='%x'\n", atomic.LoadUint64(&doneRows)+1, tableIn, key, ts, data)
+				log.Println("Query:", query)
+			}
+
+			err := session.Query(query, data, key, ts).Exec()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "ERROR: failed updating %s %s %d: %q", tableOut, key, ts, err)
+			}
+
+			doneRowsSnap := atomic.AddUint64(&doneRows, 1)
+			if doneRowsSnap%10000 == 0 {
+				doneKeysSnap := atomic.LoadUint64(&doneKeys)
+				completeness := completenessEstimate(token)
+				log.Printf("WORKING: processed %d keys, %d rows. (last token: %d, completeness estimate %.1f%%)", doneKeysSnap, doneRowsSnap, token, completeness*100)
+			}
 		}
-
-		err := session.Query(query, data, key, ts).Exec()
+		err := iter.Close()
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "ERROR: failed updating %s %s %d: %q", tableOut, key, ts, err)
-			iter.Close()
-			os.Exit(2)
+			doneKeysSnap := atomic.LoadUint64(&doneKeys)
+			doneRowsSnap := atomic.LoadUint64(&doneRows)
+			fmt.Fprintf(os.Stderr, "ERROR: failed querying %s: %q. processed %d keys, %d rows", tableIn, err, doneKeysSnap, doneRowsSnap)
 		}
-		rownum++
+		atomic.AddUint64(&doneKeys, 1)
+	}
+}
+
+func update(session *gocql.Session, ttl int, tableIn, tableOut string) {
+
+	keyItr := session.Query(fmt.Sprintf("SELECT distinct key FROM %s", tableIn)).Iter()
+
+	jobs := make(chan string, 100)
+
+	var wg sync.WaitGroup
+	for i := 0; i < *numThreads; i++ {
+		wg.Add(1)
+		go worker(i, jobs, &wg, session, *startTs, *endTs, ttl, tableIn, tableOut)
 	}
-	err := iter.Close()
+
+	var key string
+	for keyItr.Scan(&key) {
+		jobs <- key
+	}
+
+	close(jobs)
+	err := keyItr.Close()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "ERROR: failed querying %s: %q. processed %d rows", tableIn, err, rownum)
-		os.Exit(2)
+		fmt.Fprintf(os.Stderr, "ERROR: failed querying %s: %q. processed %d keys, %d rows", tableIn, err, doneKeys, doneRows)
 	}
-	fmt.Println("processed", rownum, "rows")
+
+	wg.Wait()
+	log.Printf("DONE.  Processed %d keys, %d rows", doneKeys, doneRows)
```

Comment:
might want to add 
```
if err != nil {
    os.Exit(2)
}
```
just to make sure something else calling this tool never falsely assumes everything worked

Reply:
done

Reply:
that's not important, but why not just 

```
err := keyItr.Close()
wg.Wait()
if err != nil {
    fmt.Fprintf(os.Stderr, "ERROR: failed querying %s: %q. processed %d keys, %d rows", tableIn, err, doneKeys, doneRows)
    os.Exit(2)
}
log.Printf("DONE.  Processed %d keys, %d rows", doneKeys, doneRows)
```

Reply:
you're right, that would save a line. otoh this way has the err handling right after the operation, which is nice for readability. also you can see the error without having to wait for all cassandra operations in the workers to complete.